packages/nimble-components/src/wafer-map/specs/README.md

```diff
@@ -0,0 +1,199 @@
+
+### Blazor integration
+
+Blazor support is out of scope for the current implementation.
+
+### Visual Appearance
+
+For the initial component, we are going with the original wafer map appearance. 
+![WaferMap Example](file:///Resources/wafer_map_example.png)
+
+## Implementation
+
+As mentioned above, this is not a component that we start from scratch, but a visualization that already exists which is outdated and not futureproof. 
+As a first step, we are working on the "bloat removal" which is basically the removal of unwanted but exiting features (legend, tooltip etc.) and dependencies.
+
+Original component: [wafermap](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=/libraries/wafermap&version=GBfeature/microServicesSplit)\
+Once the "bloat removal" is finished, the next step is to adapt the current 
+
+### Code submission strategy
+
+In order to make the process of submitting the re-written wafer map component easier for us and for the code reviewers as well, we propose the following bottom-up approach:
+* As an initial step, we submit a skeleton of the `nimble-wafer-map` component into the main branch. This will serve as a starting point on which the complete functionality will be built on.
+* Once the skeleton is in place, we start to work on and submit individual pieces of the functionality so it's not very overwhelming to review the code.
+
+### States
+
+We are not planning to allow the user to generate an “empty" wafer map, so we must ensure that data is provided when instantiating a new `nimble-wafer-map` component.
+The only action which we could consider as a "state transition" is the zoom functionality which was specified in the sections above.
+We expect that the end user be able to zoom in on the dies withing the wafer map and be able to zoom out by using the mouse wheel.
+Also, the user must be allowed to pan the zoomed in wafer map into any direction of its XY axis by holding the left mouse button and moving the mouse cursor around.
+
+Another interaction, which must be part of the component is the ability to identify the location of the mouse pointer on the wafer map die area.
+
+### Globalization
+
+Currently there will be no string data presented by this component. Might change in the future, but for now it will contain only numerical data.
+
+### Security
+
+N/A
+
+### Performance
+
+As mentioned in the "Risks" section, the biggest challenge from the performance standpoint is the representation of large number of datapoints and interacting with these datapoints in the browser.
+Currently the only thing we would like to achieve is to measure (benchmark) the time it takes a `nimble-wafer-map` component to be loaded with various number of datapoints.
+
+### Dependencies
+
```

Comment:
It would be good to confirm that the new dependency on `d3` doesn't bloat other clients. I would hope that:
1. clients that don't use the wafer map will not pay any bundle size or load time cost (It's ok if installing Nimble causes d3 to be installed, but don't want it to impact built apps)
2. clients that do use the wafer map won't have to download the entire library if we only use a small part of it

If we consume d3 using the ES `import` statement then this should all work for free.

Reply:
We import the following modules from d3 library:
- scaleBand
- ScaleBand
- ScaleLinear
- scaleLinear
- ScaleOrdinal
- scaleOrdinal
- select
- Selection
- zoom
- ZoomBehavior
- zoomIdentity
- zoomTransform
- ZoomTransform

Reply:
Currently we have the whole d3 library as dependency, but basically what we need is d3-scale, d3-zoom and d3-selection. 
I tested our current component and it builds and behaves as expected even using only these 3 as dependency instead of complete d3. 
The only downside is that the versions of these components are not aligned all the time: in d3 v:7.6.1 the following versions are used:
d3-zoom: v3
d3-scale: v4
d3-selection: v3

It's a bit more annoying to keep up to date 3 packages instead of the d3 library, however, this way we only use 3 out the complete 30 which comes with d3.

Reply:
Because it is just a few packages I think depending on them individually is the way to go. It will keep bundles small without relying on dead-code elimination features and I don't think will be a significant maintenance burden.

> If we consume d3 using the ES import statement then this should all work for free.

If we did rely on dead-code elimination and smart bundlers then this statement is true as long as we don't do builk imports like `import * as d3 from 'd3';`. We would have to be rigorous about importing only exactly what is needed, ie `import {zoom} from 'd3';` and rely on bundlers to dead-code eliminate the unused d3 exports.

Because we have relatively few packages that we are depending on from d3 I would vote for just using the separate packages but I'm open to either approach.

Reply:
Thanks, that's really useful background. It looks like d3 is using semantic versioning so the version mismatch might not be a problem; that just means that there have been more revisions of the `d3-scale` package than the others, not necessarily that they can't work together. Hopefully npm can just resolve any compatible differences in their sub-dependencies and show us an error if they are incompatible.

So I'd vote to depend on just the smaller packages, not all of d3. That reduces the exposure to security vulnerabilities and package bloat. I agree it's a bit of extra work to keep them up to date but long term we can automate that so it shouldn't be much of a burden.

And regarding my original point about eliminating bundle cost of d3 for clients not using the wafer map, I think the fact that you `import` d3 classes (rather than including a d3.js script on the page for example) should solve that issue.

Reply:
Looks like Milan and I replied at the same time. Thankfully we gave the same advice 😮‍💨